feat: delete ISO even if already attached

johnrichardrinehart opened the pull request:
## Description

It's currently impossible to change the URL of an ISO resource that's attached to an instance. This PR allows for the deletion of an ISO that's attached to an instance at deletion-time.

### Code/Logical Structure

1. Try do delete the ISO
2. If deletion errors, then we check to see if the error message contains information about attachment to an instance (only the IP address, the error message doesn't contain an instance ID)
3. If we detect an IP address of an instance then we try to associated the IP address with the instance ID
4. If we determine the instance ID then we try to detach the ISO from the instance (waiting until completion)
5. If detachment is successful, then we recursively call `resourceVultrIsoDelete` to attempt to delete again.

## Related Issues
#128

### Checklist:

* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [x] Have you linted your code locally prior to submission?
* [x] Have you successfully ran tests with your changes locally? **Yes, but I have written no unit tests. I tested a few times against my `terraform` infrastructure.**

johnrichardrinehart committed:
feat: delete ISO even if already attached (useful for 'hot-swapping'/changing attached ISO)

johnrichardrinehart committed:
chore: remove unused printf argument

ddymko commented:
@johnrichardrinehart Thanks for submitting this - I'll try to to get this by end of the week if not I'll have a review for ya by monday

johnrichardrinehart commented:
@ddymko Absolutely no rush. Thanks for the ping.

ddymko commented on vultr/resource_vultr_iso_private.go:
```diff
@@ -107,7 +109,71 @@ func resourceVultrIsoDelete(ctx context.Context, d *schema.ResourceData, meta in
+			return diag.Errorf("error destroying ISO %s : failed to parse IP to which ISO is attached", d.Id())
+		}
+
+		m := govultr.ListOptions{
```
I feel like this section could be simplified a bit. What I would do is condense the two for loops you have into a single loop.

Also for the recursive call that could be replaced with just a iso.Delete call. If the wait succeeds with than the delete could be called. If the delete call fails for whatever reason just return an error.

```go
		// default is 100 instances
		options := &govultr.ListOptions{}
		for {
			instances, meta, err := client.Instance.List(ctx, options)
			if err != nil {
				return diag.Errorf("error destroying ISO %s : failed to list instances for detaching ISO", d.Id(), err)
			}

			for _, instance := range instances {
				if instance.MainIP == ip {
					if err := client.Instance.DetachISO(ctx, instance.ID); err != nil {
						return diag.Errorf("error destroying ISO %s : failed to detach from instances %s : %s", d.Id(), instance.ID, err)
					}

					if _, err := waitForIsoDetached(ctx, instance.ID, "ready", []string{"isomounted"}, "status", meta); err != nil {
						return diag.Errorf("error while waiting for ISO %s to detach from instance %s: %s", d.Id(), instance.ID, err)
					}

					if err := client.ISO.Delete(ctx, d.Id()); err != nil {
						return diag.Errorf("Error deleting ISO %s : %v", d.Id(), err)
					}
				}

				if meta.Links.Next == "" {
					break
				}
				options.Cursor = meta.Links.Next
			}
		}
	}
```

johnrichardrinehart replied:
Ah, thanks! I was looking for a nice way to loop through once. Super clean! I had to modify the above a little bit, since the `break` still lands in the infinite loop.

Please see https://github.com/vultr/terraform-provider-vultr/pull/131/commits/14e4628e6c817f068290584cfb14a2e0fa01c07a

ddymko replied:
Ah yeah! my conditional  for the `meta.links.Next` was in the wrong loop.

ddymko commented on vultr/resource_vultr_iso_private.go:
```diff
@@ -107,7 +109,71 @@ func resourceVultrIsoDelete(ctx context.Context, d *schema.ResourceData, meta in
+		}
+
+		errR := strings.NewReader(err.Error())
+		jsonErr := json.NewDecoder(errR).Decode(&attachedErr)
```
Could change this to json.Unmarshal 

```go
	var attachedErr struct {
		Error  string `json:"error"`
		Status int    `json:"status"`
	}

	if err := json.Unmarshal([]byte(e.Error()), &attachedErr); err != nil {
		fmt.Println("error")
	}
```

johnrichardrinehart replied:
Yeah, that looks better than what I had.

Please see https://github.com/vultr/terraform-provider-vultr/pull/131/commits/0abc7c92d50e36870cd64e2bdf2461f15b58ef69

johnrichardrinehart committed:
refactor: remove recursive call to resourceVultrIsoDelete

refactor: use single loop for instance association with ISO Id

johnrichardrinehart committed:
refactor: changed Decode() to Unmarshal() for ISO delete error parsing

refactor: made error messages more consistent

ddymko commented on vultr/resource_vultr_iso_private.go:
```diff
@@ -107,7 +109,57 @@ func resourceVultrIsoDelete(ctx context.Context, d *schema.ResourceData, meta in
+			}
+
+			// no more instances to check
+			if options.Cursor == "" {
```
I would keep this in line with how we are doing the paging loops in other places by keeping it as 


```
	if meta.Links.Next == "" {
				break
			}
	options.Cursor = meta.Links.Next
```

and moving 

https://github.com/vultr/terraform-provider-vultr/pull/131/files#diff-9440d6827d830f384c9118b4fc75b2aea56b1c826f826fc780dbef8c208831f5R134

above the initial for loop as `options := &govultr.ListOptions{}`

johnrichardrinehart replied:
Oh, duh. 1 sec!

johnrichardrinehart replied:
Addressed in https://github.com/vultr/terraform-provider-vultr/pull/131/commits/3421f2452245d4b21281ac740bd8f7f5402c40d8 .

johnrichardrinehart committed:
fix: move options for Instance.List outside of the for loop and make the break condition clearer semantically

ddymko commented on vultr/resource_vultr_iso_private.go:
```diff
@@ -107,7 +109,57 @@ func resourceVultrIsoDelete(ctx context.Context, d *schema.ResourceData, meta in
+		// default is 100 instances
+		var options govultr.ListOptions
+		for {
+			instances, meta, err := client.Instance.List(ctx, &options)
```
This `meta` should be named something different since it overrides the meta that is passed into this function. Which causes issues when it gets handed off on line 146.

Other than this everything looks good. I am running through some tests locally now

johnrichardrinehart replied:
I missed that, thanks. The use of `interface{}` kind of masked this issue.

Please see https://github.com/vultr/terraform-provider-vultr/pull/131/commits/b8f03f2f328fae9355b64a224f7dd242acbf43f0 .

ddymko commented on vultr/resource_vultr_iso_private.go:
```diff
@@ -107,7 +109,57 @@ func resourceVultrIsoDelete(ctx context.Context, d *schema.ResourceData, meta in
+					if err != nil {
+						return diag.Errorf("error deleting ISO %s: failed to wait for ISO to detach from instance %s: %s", d.Id(), instance.ID, err)
+					}
+					if err = client.ISO.Delete(ctx, d.Id()); err != nil {
```
Also want to add a `return nil` after this conditional. 


```go
if err = client.ISO.Delete(ctx, d.Id()); err != nil {
	return diag.Errorf("error deleting ISO %s: failed to delete ISO: %s", d.Id(), err)
}
return nil
```

this way we can break out of the for loops since everything has been successfully detached + deleted

johnrichardrinehart replied:
For sure. Good catch.

Please see https://github.com/vultr/terraform-provider-vultr/pull/131/commits/b8f03f2f328fae9355b64a224f7dd242acbf43f0 .

johnrichardrinehart committed:
fix: rename meta from API response to responseMeta to avoid shadowing

fix: return from resourceVultrIsoDelete as soon as ISO is detached and
deleted successfully

ddymko reviewed (approved):
Tested locally looks good

ddymko commented:
@johnrichardrinehart Thanks for taking the time to implement this. We have some other features we want to release along side this. We are going to tag this in a 2.3.0 milestone and plan to release next week

johnrichardrinehart commented:
@ddymko Thanks for taking the time to work with me on a review. I'm much happier since we made the changes you suggested. I still don't like my solution for detecting if it's still attached (https://github.com/vultr/terraform-provider-vultr/pull/131/files#diff-9440d6827d830f384c9118b4fc75b2aea56b1c826f826fc780dbef8c208831f5R122). But, without error constants/values it's not going to look great.

I'm excited for the release and I'll be using this feature myself. Hopefully it benefits someone else :) .
